### PR TITLE
Allow customization of default odbc version

### PR DESCRIPTION
**Pull request by jabbera**
Fixes https://github.com/mkleehammer/pyodbc/issues/1277

My cpp is VERY rusty. Let me know if there is a better solution for this.

### CODE REVIEW

**Comment by jabbera** [comment]
Example code that fails with 
```
Traceback (most recent call last):
  File "/mnt/z/research_hub/work/test.py", line 21, in <module>
    connection.execute(table)
pyodbc.ProgrammingError: ('42S01', "[42S01] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]There is already an object named '#TemporaryTable' in the database. (2714) (SQLExecDirectW)")
```

without this PR and passes with it and 3.8 set.

```
import pyodbc
# pyodbc.odbcversion = "3.8" # Exception above is generated without this patch

table = """
CREATE TABLE #TemporaryTable
(
    Col1 int,
    Col2 varchar(10)
);
"""

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
```

**Comment by gordthompson** [comment]
What I notice about the repro code is that it seems to assume that pyodbc explicitly closes a connection created by a context manager. That is not the case, ref: https://github.com/mkleehammer/pyodbc/issues/48

While I can reproduce the error using the original repro code (with pyodbc 4.0.39), changing it to

```python
table = """
CREATE TABLE #TemporaryTable
(
    Col1 int,
    Col2 varchar(10)
);
"""

connection = pyodbc.connect(connection_string)
connection.execute(table)
connection.close()

connection = pyodbc.connect(connection_string)
connection.execute(table)
connection.close()

connection = pyodbc.connect(connection_string)
connection.execute(table)
connection.close()

connection = pyodbc.connect(connection_string)
connection.execute(table)
connection.close()

connection = pyodbc.connect(connection_string)
connection.execute(table)
connection.close()
```

does not reproduce the error. The difference is that the original repro code results in the `SQLEndTran` call being immediately followed by a `SQLAllocHandle` call for the "new" connection

```
[ODBC][8739][1696267076.950987][SQLEndTran.c][566]
        Exit:[SQL_SUCCESS]
[ODBC][8739][1696267076.951042][SQLAllocHandle.c][395]
        Entry:
            Handle Type = 2
            Input Handle = 0x141da00
        UNICODE Using encoding ASCII 'UTF-8' and UNICODE 'UTF16LE'
```

while explicitly closing the connection adds `SQLDisconnect` and `SQLFreeHandle` calls prior to the new `SQLAllocHandle`

```
[ODBC][9042][1696267851.430281][SQLEndTran.c][566]
        Exit:[SQL_SUCCESS]
[ODBC][9042][1696267851.430313][SQLDisconnect.c][208]
        Entry:
            Connection = 0x10a9d00
[ODBC][9042][1696267851.430344][SQLDisconnect.c][302]
        Exit:[SQL_SUCCESS]
[ODBC][9042][1696267851.430374][SQLFreeHandle.c][290]
        Entry:
            Handle Type = 2
            Input Handle = 0x10a9d00
[ODBC][9042][1696267851.430408][SQLFreeHandle.c][339]
        Exit:[SQL_SUCCESS]
[ODBC][9042][1696267851.430458][SQLAllocHandle.c][395]
        Entry:
            Handle Type = 2
            Input Handle = 0x114c0d0
        UNICODE Using encoding ASCII 'UTF-8' and UNICODE 'UTF16LE'
```

I also notice that the original repro code fails on the third context manager, not the second, which makes me suspect that "closing" a connection by clobbering the existing `connection` variable with another `.connect()` call might not behave as one might assume.

TL;DR - I'm just a bit confused. Perhaps @v-chojas might have some insights on this.

**Comment by jabbera** [comment]
I'll certainly wait on the expert to comment but I'll add my 2c here also. Even if the particular issue demoed can be resolved by using the connection correctly I'd think we would want the ability for the connection to be reset when it gets added back into the pool which only happens with 3.8. Some of the items it does are not covered by the endtran call. Additionally if a connection leaks (not calling close) accidentally you won't end up with a dirty connection. 

Also, while I demoed this issue using pyodbc directly I stumbled across it using sqlalchemy with the nullpool because we prefer unixodbc pooling. I would be shocked if they were missing the close call, but I'll dig into it.

**Comment by jabbera** [comment]
@gordthompson now I'm really confused. With 4.0.39 this produces the error even though the connection is being closed:

```
with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
connection.close()

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
connection.close()

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
connection.close()

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
connection.close()

with pyodbc.connect(connection_string) as connection:
    connection.execute(table)
connection.close()

```

**Comment by jabbera** [comment]
@gordthompson I think you have a bug in your test case. Because autocommit is not enabled close in your case rolls back the transaction that creates the temporary table. If you turn on autocommit or add a commit statement the bug reproduces fine:

```
for i in range(20):
    connection = pyodbc.connect(connection_string)
    connection.execute(table)
    connection.commit()
    connection.close()
```

**Review by gordthompson — COMMENTED** [review]
Looks good to me.

**Comment by mkleehammer** [comment]
Thanks.  In the future, we might want to add an error message if the user tries something unsupported like "3.9".
